text/3349-cfg-os-version.md

```diff
@@ -0,0 +1,142 @@
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rust targets are often thought of as monoliths.  The thought is that if you compile a binary for that target, that binary should be able to run on any system that fits that target's description. However, this is not actually true.  For example, when compiling for `x86_64-pc-windows-msvc` and linking with the standard library, my binary has implicitly taken a dependency on a set of APIs that Windows exposes for certain functionality.  If I try to run my binary on older systems that do not have those APIs, then my binary will fail to run.  When compiling for a certain target, you are therefore declaring a dependency on a minimum target API version that you rely on for your binary to run.
+
+By default, the standard library uses a sensible minimum API version.  For example, for `x86_64-pc-windows-msvc` the minimum API version is "6.1.7600" which corresponds to Windows 7.  However, there are good reasons why you might want control of how your code is compiled depending on what the minimum API version is set as.  For instance, if you want to:
+
+* set your crate's minimum API version higher than that of the standard library.
+* change certain implementation details of your crate depending on what a downstream user sets their minimum API version to be.
+* have some sensible compiler error if users of your crate require a lower minimum API version that you require.
+
+In cases like these you can use the `os_version_min` and `os_version_range` predicates to specify the minimum API levels of various parts of the operating system.  For example:
+
+* `os_version_min(“windows”, <string>)` would test the [minimum build version](https://gaijin.at/en/infos/windows-version-numbers) of Windows.
+* `os_version_min(“libc”, <string>)` would test the version of libc in use.
+* `os_version_min(“kernel”, <string>)` would test the version of the kernel in use.
+
+Let’s use `os_version_min(“windows”, …)` as an example.  It should be clear how this example would be extended to the other `cfg` predicates. The predicate allows you to conditionally compile code based on the set minimum API version. For example an implementation of mutex locking on Windows might look like this:
+
+```rust
+pub unsafe fn unlock(&self) {
+    *self.held.get() = false;
+    if cfg!(os_version_min(“windows”, "6.0.6000") { // API version greater than Vista
+        c::ReleaseSRWLockExclusive(raw(self)) // Use the optimized ReleaseSRWLockExclusive routine
+    } else {
+        (*self.remutex()).unlock()  // Fall back to an alternative that works on older Windows versions
+    }
+}
+```
+
+Crate authors can set the API requirements of their Cargo configuration file under the  [`target key`](https://doc.rust-lang.org/cargo/reference/config.html#target) like so (suggested variable name/syntax only):
+
+```toml
+[target.x86_64-pc-windows-msvc]
+min_os_version_windows = "6.0.6000" # Vista
```

Comment:
It would be nice to have specified the cases where raising `min_os_version_*` in a crate is a breaking change (e.g. if the standard library has raised theirs first, it probably isn't, but in other cases it may be).